python/pyspark/sql/dataframe.py

```diff
@@ -2153,6 +2153,59 @@ def transform(self, func):
                                               "should have been DataFrame." % type(result)
         return result
 
+    @since(3.1)
+    def sameSemantics(self, other):
+        """
+        Returns `True` when the logical query plans inside both :class:`DataFrame`\\s are equal and
+        therefore return same results.
+
+        .. note:: The equality comparison here is simplified by tolerating the cosmetic differences
+            such as attribute names.
+
+        .. note::This API can compare both :class:`DataFrame`\\s very fast but can still return
+            `False` on the :class:`DataFrame` that return the same results, for instance, from
+            different plans. Such false negative semantic can be useful when caching as an example.
+
+        >>> df1 = spark.range(100)
+        >>> df2 = spark.range(100)
+        >>> df3 = spark.range(100)
+        >>> df4 = spark.range(100)
+        >>> df1.withColumn("col1", df1.id * 2).sameSemantics(df2.withColumn("col1", df2.id * 2))
+        True
+        >>> df1.withColumn("col1", df1.id * 2).sameSemantics(df3.withColumn("col1", df3.id + 2))
+        False
+        >>> df1.withColumn("col1", df1.id * 2).sameSemantics(df4.withColumn("col0", df4.id * 2))
+        True
+        """
+        if not isinstance(other, DataFrame):
+            raise ValueError("other parameter should be of DataFrame; however, got %s"
+                             % type(other))
+        return self._jdf.sameSemantics(other._jdf)
+
+    @since(3.1)
+    def semanticHash(self):
+        """
+        Returns a hash code of the logical query plan against this :class:`DataFrame`.
+
+        .. note:: Unlike the standard hash code, the hash is calculated against the query plan
+            simplified by tolerating the cosmetic differences such as attribute names.
+
+        >>> df1 = spark.range(100)
+        >>> df2 = spark.range(100)
+        >>> df3 = spark.range(100)
+        >>> df4 = spark.range(100)
+        >>> df1.withColumn("col1", df1.id * 2).semanticHash() == \
+            df2.withColumn("col1", df2.id * 2).semanticHash()
+        True
+        >>> df1.withColumn("col1", df1.id * 2).semanticHash() == \
+            df3.withColumn("col1", df3.id + 2).semanticHash()
+        False
```

Comment:
```
Failed example:
    df1.withColumn("col1", df1.id * 2).semanticHash() ==             df3.withColumn("col1", df3.id + 2).semanticHash()
Differences (ndiff with -expected +actual):
    - False
    + True
```
Now we have another unexpected result. (Note L2176 passed, which is expected.)

Reply:
More tests:
```
>>> df1=spark.range(100)
>>> df2=spark.range(100)
>>> df3=spark.range(100)
>>> df11=df1.withColumn("col1", df1.id +1)
>>> df21=df2.withColumn("col1", df2.id -1)
>>> df31=df3.withColumn("col1", df3.id *2)
>>> df32=df3.withColumn("col1", df3.id +2)
>>> df33=df3.withColumn("col1", df3.id /2)
>>> df34=df3.withColumn("col1", df3.id -2)

>>> df11.semanticHash()
1855039936
>>> df21.semanticHash()
1855039936

>>> df31.semanticHash()
-1719131362
>>> df32.semanticHash()
-1719131362

>>> df32.semanticHash()
-1719131362
>>> df34.semanticHash()
-706037631

Reply:
Same behavior for dataframe from `spark.read.load()`
```
>>> df4=spark.read.load(csv_file_path, format="csv", inferSchema="true", header="true")
>>> df4.schema
StructType(List(StructField(bool_col,BooleanType,true),StructField(float_col,DoubleType,true),StructField(double_col,DoubleType,true),StructField(int_col,IntegerType,true),StructField(long_col,IntegerType,true)))
>>> df4.withColumn("col1", df4.int_col *2).semanticHash()
-1746346451
>>> df4.withColumn("col1", df4.int_col +2).semanticHash()
-1746346451

Reply:
@liangz1 I think you should check correctness via df.sameSemantics, don't via hash, hash may cause collisions issue.

Reply:
I am concerned that the hash collision rate is too high, which could be contradictory to people's expectations. Typically, if two messages (in this case, dataframes) have subtle differences, the hash should be different. Do you think we should add a warning in the docstring?
@HyukjinKwon @WeichenXu123 @mengxr

Reply:
Okay, this seems an issue in Scala. I will open a PR.

Reply:
Thanks! Could you share the link to the PR or issue?

Reply:
Sure, PR: https://github.com/apache/spark/pull/27601 and the JIRA: https://issues.apache.org/jira/browse/SPARK-30847